052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,62 @@
+
+# Kubernetes Server Side Apply
+
+Make the Operators use [Kubernetes Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply). This will enable third parties (e.g. Kyverno, Kopf) to have their own labels/annotations on Strimzi owned Kubernetes Resources without Strimzi attempting to remove them.
+
+NOTE: Strimzi will still have ownership over the fields it wants.
+
+## Current situation
+
+As per [Issue 6938](https://github.com/strimzi/strimzi-kafka-operator/issues/6938) third party Kubernetes operators are having a negative interaction with the existing way Strimzi does Kubernetes applies.
+An example is [Kyverno](https://kyverno.io/) has a policy in a K8s cluster to add a label to all resources with a MutatingWebhookConfiguration, Strimzi detects this label as a resource diff and removes it as it doesn't own it / not included in [hard-coded list of ignored labels](https://github.com/strimzi/strimzi-kafka-operator/blob/c3522cf4b17004004a676854d37ba01bb9a44800/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetDiff.java#LL28C34-L28C49). This triggers the MutatingWebhookConfiguration which re-applies the label and the loop continues.
+
+## Motivation
+
+Once this change is made (which is relatively simple to implement), it will allow third party tools within the cluster to interact with Strimzi owned resources without looping and Strimzi will be operating as if it never knew about them.
+
+## Proposal
+
+This proposal makes Kubernetes keep track of each field's "owner" and any changes submitted to those fields will be compared to the remote state before being applied. 
+
+This would let the Kubernetes apiserver handle the three-way diff, updates etc, and allow other controllers, webhooks etc to touch fields, labels, annotations etc the operator doesn't care about.
+
+It appears as simple as modifying [this line](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/AbstractNamespacedResourceOperator.java#LL263C88-L263C102).
+From:
+
+```
+....patch(PatchContext.of(PatchType.JSON)...
+```
+
+To:
+
+```
+....patch(PatchContext.of(PatchType.SERVER_SIDE_APPLY)...
+```
+
+According to the [package being used](https://github.com/fabric8io/kubernetes-client/blob/v6.5.1/doc/CHEATSHEET.md#server-side-apply).
+
+NOTE: The version of fabric8io being used is [6.5.1](https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/pom.xml#L106) which [contains Server Side Apply](https://github.com/fabric8io/kubernetes-client/blob/v6.5.1/doc/CHEATSHEET.md#server-side-apply).
+
+It is likely that other parts of the codebase will need modifying in terms of label/annotation discovery so that they only check labels/annotations that are owned by themselves.
```

Comment:
I think this is one of the big ones: https://github.com/strimzi/strimzi-kafka-operator/blob/c3522cf4b17004004a676854d37ba01bb9a44800/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ResourceDiff.java#L46-L78

Instead of it generating a diff of everything that's different instead of a specified list of ignorable paths, it should be generating a diff of all of the paths specified in desired (and ignore everything that wasn't specified)

Reply:
Then all of the `ignorablePaths` bits throughout the codebase can be removed, as can bits like https://github.com/strimzi/strimzi-kafka-operator/blob/18d76bfabcfb9e91c71f9afda60b9dd880797f02/operator-common/src/main/java/io/strimzi/operator/common/operator/resource/ServiceOperator.java#L119

(Currently it's adding in a specific set of annotations from the live cluster into the desired state, so they aren't then diffed etc. With the new approach that's not necessary, as the diff won't care about unknown annotations)

Reply:
I'm not sure this is the case. Not setting something is not the same as not caring about how it is configured. So I think you can get rid of the ignorable paths, but not completely sure you can get rid of the diff completely. The diff also saves resources so it makes one wonder how would that look with server-side apply ... is calling server-side apply all the time a good choice even without any change?

Reply:
Removing the diff completely has never been suggested.
What needs to happen is a diff on purely the fields that Strimzi have control over and would want to reconcile on, which means we would not need `ignorablePaths` as they shouldn't be included in the diff to start with, i.e. we don't want to trigger a reconcile on a diff generated by an annotation managed by a third party.

This will then allow the diff to work hand-in-hand with Server Side Apply.

What do you mean by:
> The diff also saves resources

?

Reply:
> What do you mean by:
> 
> > The diff also saves resources
> 
> ?

Thanks to the diff, we do not trigger unnecessary update calls to the Kubernetes API because we know that nothing changed and there is no need to patch it.